### PR TITLE
Support for modifiers in KeyboardHandler

### PR DESCRIPTION
**Pull request by entrhopi**
AFAIK it has only 2 minor issues atm:
- number keys do not respect keyboard layout (it always uses US layout)
- holding shift for too long will kind of lock it for the next key (does not impact normal writing)

### CODE REVIEW

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (dlamkins) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: Can we mark this with a `//TODO: ` or something to indicate that we need to review in the future?  I'll create an issue to track it since certainly we'll want to aim for better keyboard layout support in the future, but I think this could be a fine solution for now.  Better than no support.

**Review by dlamkins — CHANGES_REQUESTED** [review]

**Comment by greaka** [comment]
Just leaving this here for the future: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-tounicode

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (entrhopi) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: I've committed a patch handling all the issues I've found so far.
But I'll mark potential places in the future.

**Comment by entrhopi** [comment]
> Just leaving this here for the future: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-tounicode

Isn't the ToUnicodeEx we're using here 'better', because it handles keyboard layouts?
My problem was that the shift-modifier wasn't working and the `ToUpper` way of capitalizing chars was not.

We'll probably have to rework the entire keyboard handling anyway as I would consider my patch only a temporary workaround to get basic things working now.

**Comment by agaertner** [comment]
> Just leaving this here for the future: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-tounicode

Also this https://stackoverflow.com/questions/17386766/how-do-i-get-a-char-obtained-by-holding-shift

Glad I was able to help with this.

I know TacO uses the ToASCIIEx function but I just saw the ToUnicodeEx function also takes an array where you could pass the shift key pressed state in.

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (agaertner) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: Worth trying ToUnicode to reflect the old implementation.

Comment (entrhopi) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: I would like to revisit the whole keyboard handling in the near future but I think it's more important to have a quick solution for now that is working since the shift key is't working at all and somehow important for the functionality of at least the KillProof module.

Comment (agaertner) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: I was trying out your fork and it appears sometimes the bug that the shift toggle becomes active at the second or third letter still happens. There seems to be still something off with the state.

Comment (entrhopi) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: I think I've found what's causing that. I'm currently reworking the whole function to be more rubust and use `ToUnicodeEx`.
Please consider my PR as on-hold until the next commit.

Comment (entrhopi) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: I just noticed that accent and diacritic characters aren't working in either implementation so I'll fix that too.

Comment (entrhopi) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: So I've made some modifications but as @greaka pointed out setting the key state for shift manually is not the most elegant solution and I agree.
However, the problem seems that using `GetKeyboardState` and passing it directly to `ToUnicode` isn't working because it seems to set the low-order bit for the keys which is for 'toggle keys' whereas `ToUnicode` expects the high-oder bit set.
Here is what I've tried:
```cs
byte[] keyState = new byte[256];
GetKeyboardState(keyState);
ToUnicode(virtKeyCode, scanCode, keyState, output, (int)5, (uint)0);
```
Even getting some key states before calling `GetKeyboardState` as mentioned here ( https://stackoverflow.com/a/53713024 ) does not work (for me at least).

Comment (entrhopi) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: New insights:
Adding `Keys.(Left|Right)Shift` and `Keys.(Left|Right)Control` to `_hookIgnoredKeys` is necessary for the systems `User32.dll` to know of the states of these keys.
`GetKeyboardState` will then work if a `GetKeyState` is called before as stated in the SO post above.

This leaves us with all modified keys working.

Dead key implementation still pending.

Comment (entrhopi) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: Dead-key characters are now implemented too.

I think we're golden now.

Comment (agaertner) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: Tested. The bug that caps begins to apply to the second or third letter appears to be gone. 🚀

However, the current implementation seems to be blocking the key-combos in the [TextInputBase](https://github.com/blish-hud/Blish-HUD/blob/0e82ee8668b1e8fcb835795c4735e1ab360947e9/Blish%20HUD/Controls/TextInputBase.cs#L214-L243) or that control base has to be updated. Mainly I tried Ctrl+A (select all), Ctrl+C (copy), Ctrl+V (paste)

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (agaertner) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: redundant ``= false``

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (agaertner) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: ``ToUnicode``

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (agaertner) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: can be made private as all references appear to be in TypedInputUtil.cs.

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (agaertner) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: internal scope might be more appropriate since we are dealing with keyboard state

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (agaertner) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: redundant ``= 0``

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (agaertner) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: *apparently

**Review by agaertner — APPROVED** [review]
Tested. Everything appears to be working now. 👍 
Consider scopes and typos before merging.

**Comment by entrhopi** [comment]
> Tested. Everything appears to be working now. 👍
> Consider scopes and typos before merging.

Thank you, will refactor.

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (agaertner) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: empty doc tags can be removed.

File: Blish HUD/Controls/TextInputBase.cs
Comment (agaertner) [inline Blish HUD/Controls/TextInputBase.cs:563]: It might be better to use
```csharp
(!clipboardResult.IsCompleted)
```
instead in a ContinueWith since IsCompleted covers 
Canceled, Faulted and RanToCompletion according to [docs](https://docs.microsoft.com/de-de/dotnet/api/system.threading.tasks.task.iscompleted?redirectedfrom=MSDN&view=net-5.0#System_Threading_Tasks_Task_IsCompleted).

File: Blish HUD/Controls/TextInputBase.cs
Comment (agaertner) [inline Blish HUD/Controls/TextInputBase.cs:578]: It might be better to use
```csharp
(!clipboardResult.IsCompleted)
```
instead in a ContinueWith since IsCompleted covers 
Canceled, Faulted and RanToCompletion according to [docs](https://docs.microsoft.com/de-de/dotnet/api/system.threading.tasks.task.iscompleted?redirectedfrom=MSDN&view=net-5.0#System_Threading_Tasks_Task_IsCompleted).

File: Blish HUD/Controls/TextInputBase.cs
Comment (agaertner) [inline Blish HUD/Controls/TextInputBase.cs:581]: redundant parenthesis

Comment (agaertner) [inline Blish HUD/Controls/TextInputBase.cs:581]: try reduce redundant paranthesis in single-line nested IF-statements using indent instead. Might look better.

**Review by greaka — APPROVED** [review]
I am against the open proposed changes by Andy and would merge this as is after hearing back from the testers.

File: Blish HUD/Controls/TextInputBase.cs
Comment (dlamkins) [reply Blish HUD/Controls/TextInputBase.cs:563]: This wasn't part of the original PR - it was just a formatting thing that got included.  We also only want to log an entry if it strictly faulted.  If it gets canceled or otherwise, no reason to log it because there won't be an exception included.

File: Blish HUD/Controls/TextInputBase.cs
Comment (dlamkins) [reply Blish HUD/Controls/TextInputBase.cs:581]: Not really part of this PR - appears to be an artifact of some minor formatting that got applied to the whole file.  👍

File: Blish HUD/GameServices/Input/Keyboard/KeyboardHandler.cs
Comment (dlamkins) [inline Blish HUD/GameServices/Input/Keyboard/KeyboardHandler.cs]: Since this function is accessible outside of the class.  Let's keep the name as `VkCodeToString` (capital `V`).

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (dlamkins) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs]: These are walking and talking like consts.  Let's make 'em consts and place them near the top of the file (either before or after the DLLImports is fine).

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (dlamkins) [inline Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs:12]: I might have missed it.  Was there any reason in particular we decided to go for `ToUnicode` instead of `ToUnicodeEx` for this?

**Review by dlamkins — CHANGES_REQUESTED** [review]

**Comment by dlamkins** [comment]
This is looking absolutely fantastic.  I posted my final thoughts prior to approval.

Assuming Ni has no issues when he tests it, I suspect we're nearly ready to merge!  🎉

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (dlamkins) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs:12]: (this is not a blocker)

Comment (agaertner) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs:12]: It may have had something todo with letting the OS handle the keyboard layout and not ourselfs since ``ToUnicodeEx`` expects a local identifier of a keyboard layout used in translating the code ie. it appears to be just a means to force a different keyboard layout used in translation if you'd want that. ``ToUnicode`` uses the current active one.

File: Blish HUD/GameServices/Input/Keyboard/KeyboardHandler.cs
Comment (agaertner) [reply Blish HUD/GameServices/Input/Keyboard/KeyboardHandler.cs]: Good catch.

File: Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs
Comment (entrhopi) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs:12]: Yes. @greaka and I were discussing this. Since we don't need to inject our own layout at any point we ditched the `ToUnicodeEx`. Reading the current layout and injecting it would have been unnecessary and redundant code.

Comment (greaka) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs:12]: `ToUnicodeEx` requires us to handle keyboard layouts ourselves which `ToUnicode` does on its own. When a user switches layouts, we either have to poll layouts again every update or we don't support switching layouts. We just don't want or need the control `ToUnicodeEx` provides

Comment (dlamkins) [reply Blish HUD/GameServices/Input/Keyboard/TypedInputUtil.cs:12]: Alright, thanks for the explanations!  That makes perfect sense.

**Review by dlamkins — APPROVED** [review]

**Comment by dlamkins** [comment]
Closes #5!

**Pull request merged**
